software/19.Socket.io/README.md

```diff
@@ -0,0 +1,235 @@
+- [Base Print](https://www.typescriptlang.org/docs/handbook/typescript-from-scratch.html)
+- [Emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
+
+## Step 05 - Get Browser Data
+
+Now that we will get the data, we will send it from our server in index.ts.
+
+The objective of this step is to know when the user presses the button on your html page.
+At this point you will add the content of the input into your terminal.
+
+It is time to get information on your input.
+
+First step: create two const variable:
+* form_data: get element form by id on your html page.
+* input_data: get element input by id on your html page.
+
+Second step:
+* Listen to the event submit button.
+* Create a `message` variable that gets the data from your input.
+* Print content message with console.log().
+* Call emit_data to sent `message`.
+* Established input in Null.
+* Tips: to get value of your data_input cast your variable with `HTMLInputElement`.
+
+Go to your console on your shell and watch if you receive the message when you
+click on the button `Send`.
+
+#### Resources :
+- [Socket listener](https://socket.io/docs/v4/index.html)
+- [Cast Variable](https://www.typescripttutorial.net/typescript-tutorial/type-casting/)
+
+## Step 06 - Print Data in Your Browser Console
+
+Having our message on your shell is great but receive your message in your browser console is better.
+
+In your server `index.ts` changed the sending of sockets.
+Make sure that everyone can receive the message.
+
+* Change your socket.emit of your file `index.ts`for send socket to all clients in the current namespace except the sender.
+
+Now, if you submit message and look your console browser you must see your message.
+
+#### Resources :
+- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
+- [Type of emit Socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
+
+## Step 07 - Print Data in your Browser Html Page
+
+Now we will create a function in our `Chat` class to add the content of our input to our div.
+This will allow us to see the messages directly in our HTML page.
+
+For this, we will create a function `append_data` on your class Chat.
+First step, create const variable:
+* html_page: get element by id of your `div`, the name of this id is `messages'.
+
+Your second step is in your function `append_data`:
+* Create variable `add_element` like element  `div`.
+* The attribute the value of this variable to the message of your parameter.
+* Append this variable on your variable `html_page`.
+
+Finally, call this function in the function `emit_data`.
+
+If you send a message, you must see in your HTML page !!!!!
+Congratulation !
+
+#### Resources :
+
+- [Release a socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
+- [Request HTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)
+- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
```

Comment:
```suggestion
Now we will create a new method in our `Chat` class to add the content of our input into our div.
This will allow us to see the messages directly in our HTML page.

To do this, we will create a method `append_data` in our class `Chat`.
First, create a constant variable:
- `html_page`: get element by his id of you `div`, his id should be is `messages'.

Second, in your method `append_data`:
- Create a variable `add_element` that should create an element `div`.
- Display the message gave as parameter in this new div.
- Append the `div` to your variable `html_page`.

Finally, call this method in the `emit_data`.

If you send a message, you should see it in your HTML page !!!
Congratulation !

#### Resources
- [Release a socket](https://socket.io/docs/v3/emit-cheatsheet/index.html)
- [Request HTML](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelector)
- [Socket Client](https://socket.io/docs/v4/client-socket-instance/)
```